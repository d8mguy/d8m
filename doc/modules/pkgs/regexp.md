## Regexp

Module regexp provides access to a subset of the functions in Go's `regexp` package, plus a d8m type `scanexpT` that can match multiple regular expressions and thereby serve as a simple lexical scanner. 

#### Type regexpT

    [regexpT: pattern]
where `pattern: string`.

    val regexpT = extend tuple(private goptn: rxpkg.Regexp, strg: string) where {
        method find = \(s: list(byte)) -> list(byte) {...}
    	method findIndex = \(s: list(byte)) -> list(integer) {...}
    	method findAllIndices = \(s: list(byte)) -> list(list(integer)) {...}
	    method split = \(s: list(byte)) -> list(list(byte)) {...}
    }

Method `find`: returns the first sequence of bytes matching the pattern, `[]` for no match. 

Method `findIndex`: returns two integers indexing start and end of the first sequence of bytes matching the pattern, `[]` for no match. 

Method `findAllIndices`: returns two integers indexing start and end of the first sequence of bytes matching the pattern, `[]` for no match. 

Method `split`: does a version of split where the separator is the regexp `self` rather than a fixed `list(byte)`. Returns the `list(byte)` between each separator. 

#### Type scanexpT

A `scanexpT` is analogous to a `regexpT` but scans multiple regexps, given in the constructor. This assumes the given strings are mutually exclusive wrt matching, ie at most one can match.

    [scanexpT: ss]
where `ss:list(string)` is a list of pattern strings which get wrapped in () and concatenated with | so as to make a single pattern that can be submatched and post-processing can find which pattern matched each part of the original.

    val scanexpT = extend tuple(strgs: list(string)) where {
        method scan = \imp(bytes: list(byte)) -> list(integer) {...}
        method done = \(bytes: list(byte), ofs: integer) -> boolean {...}
        method next = \imp(bytes: list(byte), ofs: integer) -> list(integer) {...}
        method checkWhite = \imp(bytes: list(byte), indices: list(integer)) -> integer {...}
    }
Method `scan`: runs FindAllSubmatchIndex and processes the result into a list of triples, the match indices plus index of the submatch (i.e., the index of the match in the list argument to the constructor).

Method `done`: `done` and `next` provide a way to scan a scanexpT incrementally.

Method `next`: returns a 3-elt list in the same format as scan, or an empty list if done.

Method `checkWhite`: The indices arg should have been returned from scan. Return index of the first non-whitespace byte between tokens or -1 if no such. Thus, -1 is the expected return if the tokens cover the line, except for whitespace. Whitespace is defined as space and tab characters. 