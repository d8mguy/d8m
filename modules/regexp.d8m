/*
 * provide a d8m interface to go's regexp package
 */
import go "regexp" as rxpkg

export regexpT, scanexpT

// Create a regexpT with a string representing a regexp; it holds the internal ("compiled") version of same and
// provides access methods to match entities typed as list(byte).
val regexpT = extend tuple(private goptn: rxpkg.Regexp, strg: string) where {
	method lithook = \mod(s: string) {
		goptn = rxpkg.MustCompile(s)
		strg = s
	}
	// Go Find returns []byte, with [] for no match. That's ambiguous with a pattern that matches the empty string, but
	// you probably shouldn't use such a pattern.
	method find = \(s: list(byte)) -> list(byte) { goptn.Find(s) }
	// Note that although the go documentations says findIndex returns nil if not found, it actually returns [].
	method findIndex = \(s: list(byte)) -> list(integer) { goptn.FindIndex(s) }
	// the retvals from findAllIndices will have length 2 lists inner; this is inherited from go, doesn't seem worth redoing
	method findAllIndices = \(s: list(byte)) -> list(list(integer)) { goptn.FindAllIndex(s, -1) }
	method split = \(s: list(byte)) -> list(list(byte)) {
		val matches = goptn.FindAllIndex(s, -1)
		if(matches.count == 0) [s]
		else {
			// find the negative spaces between matches to select
			val starts = [0] + matches.{this[1]}
			val ends = matches.{this[0]} + [s.count]
			[st~starts, ed~ends].xp.{ s.select(st, ed) }
		}
	}
}

// A scanexpT is analogous to a regexpT but scans multiple regexps, given in the constructor.
// This assumes the given strings are mutually exclusive wrt matching, ie at most one can match.
val scanexpT = extend tuple(private goptn: rxpkg.Regexp, strgs: list(string)) where {
    // `ss` is a list of pattern strings which get wrapped in () and concatenated with | so as to make a
    // single pattern that can be submatched and post-processing can find which pattern matched each part of the original.
    method lithook = \mod(ss: list(string)) {
        if(ss.count == 0) exit("scanexp requires at least one pattern")
        val count1 = sigma(ss.{this.count})      // get length of all strings in ss
        var concatd: list(byte) = zerolist(count1 + 3*ss.count - 1)
        var inx = 0
        val lastStrg = ss.last
        each(strg^ss) {
            concatd[inx] = '('; inx += 1
            concatd.copyFrom(inx, strg)
            inx += strg.count
            concatd[inx] = ')'; inx += 1
            if(strg != lastStrg) { concatd[inx] = '|'; inx += 1 }
        }
        goptn = rxpkg.MustCompile(cvt(concatd, string))
        strgs = ss
    }

    // runs FindAllSubmatchIndex and process the result into a list of triples, the match indices plus index of submatch.
    method scan = \imp(bytes: list(byte)) -> list(integer) {
        val scanned = goptn.FindAllSubmatchIndex(bytes, -1)
        // scanned is list(list(integer)) where each elt is length 2*(ss.count+1); the first 2 are the match indices
        // and the others should be -1,-1 except for one pair which indicates the submatch that hit.
        // Process that into triples.
        var rslt: list(integer) = zerolist(scanned.count * 3)
        each(mtch^scanned, inxRslt) {
            rslt[inxRslt*3] = mtch[0]; rslt[inxRslt*3+1] = mtch[1]
            each(i^(1...mtch.count/2)) if(mtch[i*2] != -1) { rslt[inxRslt*3+2] = i-1; break }
        }
        rslt
    }

    // done and next provide a way to scan a scanexpT incrementally.
    method done = \(bytes: list(byte), ofs: integer) -> boolean {
        goptn.FindSubmatchIndex(bytes[ofs...bytes.count]).count == 0
    }

    // returns a 3-elt list in the same format as scan, or an empty list if done
    method next = \imp(bytes: list(byte), ofs: integer) -> list(integer) {
        val match = goptn.FindSubmatchIndex(bytes[ofs...bytes.count])
        if(match.count != 0)
            each(i^(1...match.count/2)) if(match[i*2] != -1) return [ofs + match[i*2], ofs + match[i*2+1], i-1]
        []
    }

    // The indices arg should have been returned from scan. Return index of the first non-whitespace byte between tokens
    // or -1 if no such. Thus, -1 is the expected return if the tokens cover the line, except for whitespace.
    method checkWhite = \imp(bytes: list(byte), indices: list(integer)) {
        var lasttop = 0
        var inx = 0
        while(inx < indices.count) {
            val inx0 = indices[inx]
            while(lasttop < inx0) {
                unless(bytes[lasttop] == ' ' || bytes[lasttop] == '\t') return lasttop
                lasttop += 1
            }
            lasttop = indices[inx+1]
            inx += 3
        }
        -1
    }
}
