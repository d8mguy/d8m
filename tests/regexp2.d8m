// regexp2
import "regexp"
val tryRE = \(strg1, strg2: string) {
    val ptnstrgs = [",", "[a-zA-Z_][a-zA-Z_0-9]*", "\".*\"", "`.*`", "[0-9]*\\.[0-9]+", "[0-9]+"]
    val re1 = [regexp.scanexpT: ptnstrgs]
    println(re1.scan(strg1))
    val s2 = re1.scan(strg2)
    println(s2)
    println(re1.checkWhite(strg2, s2))
}
tryRE(" eiu, 388, \"ieu*;\", `SS;**8;xxxd`", "e_iu, ;**388, ` ieu*; S(S;`, xxxd")
