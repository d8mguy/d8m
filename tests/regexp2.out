[1, 4, 1, 4, 5, 0, 6, 9, 5, 9, 10, 0, 11, 18, 2, 18, 19, 0, 20, 33, 3]
[0, 4, 1, 4, 5, 0, 9, 12, 5, 12, 13, 0, 14, 27, 3, 27, 28, 0, 29, 33, 1]
6
